### PR TITLE
New matrix allocator and initializer to Level 3, _ex and Solver functions

### PR DESCRIPTION
**Pull request by NaveenElumalaiAMD**
resolves:[ lwpmlse-644](https://ontrack-internal.amd.com/browse/LWPMLSE-644)

### CODE REVIEW

File: clients/include/blas3/testing_hemm.hpp [hunk 1]
```diff
@@ -55,11 +55,12 @@ void testing_hemm_bad_arg(const Arguments& arg)
     hipblasSideMode_t side = HIPBLAS_SIDE_LEFT;
     hipblasFillMode_t uplo = HIPBLAS_FILL_MODE_LOWER;
 
-    int64_t colsA = side == HIPBLAS_SIDE_LEFT ? N : M;
+    size_t dim_A = (side == HIPBLAS_SIDE_LEFT ? N : M);
```
Comment (TorreZuk) [inline clients/include/blas3/testing_hemm.hpp]: Hmmm, think this flipped as discussed in rocBLAS.  Left would be M x M otherwise N x N.  @daineAMD  ?

Comment (daineAMD) [reply clients/include/blas3/testing_hemm.hpp]: Yes, should be swapped.

File: clients/include/blas3/testing_symm.hpp [hunk 2]
```diff
@@ -55,11 +55,12 @@ void testing_symm_bad_arg(const Arguments& arg)
     hipblasSideMode_t side = HIPBLAS_SIDE_LEFT;
     hipblasFillMode_t uplo = HIPBLAS_FILL_MODE_LOWER;
 
-    int64_t colsA = side == HIPBLAS_SIDE_LEFT ? N : M;
+    size_t dim_A = (side == HIPBLAS_SIDE_LEFT ? N : M);
```
Comment (daineAMD) [inline clients/include/blas3/testing_symm.hpp]: Same here as hemm.

File: clients/include/blas3/testing_syr2k.hpp [hunk 3]
```diff
@@ -55,11 +55,13 @@ void testing_syr2k_bad_arg(const Arguments& arg)
     hipblasOperation_t transA = HIPBLAS_OP_N;
     hipblasFillMode_t  uplo   = HIPBLAS_FILL_MODE_LOWER;
 
-    int64_t cols = transA == HIPBLAS_OP_N ? K : N;
+    size_t rows = (transA != HIPBLAS_OP_N ? std::max(K, int64_t(1)) : N);
+    size_t cols = (transA == HIPBLAS_OP_N ? std::max(K, int64_t(1)) : N);
```
Comment (daineAMD) [inline clients/include/blas3/testing_syr2k.hpp:59]: May want to have these in the same order between files; having rows first sometimes and cols first other times is more likely to cause a coding error imo.

File: clients/include/blas3/testing_trmm.hpp [hunk 4]
```diff
@@ -79,16 +78,19 @@ void testing_trmm_bad_arg(const Arguments& arg)
             zero = zero_d;
         }
 
-        size_t A_size = lda * K;
-        size_t B_size = ldb * N;
+        int64_t K = side == HIPBLAS_SIDE_LEFT ? M : N;
```
Comment (daineAMD) [inline clients/include/blas3/testing_trmm.hpp:81]: Trmm has this correct here.

File: clients/include/blas3/testing_trsm_batched.hpp [hunk 5]
```diff
@@ -362,42 +361,45 @@ void testing_trsm_batched(const Arguments& arg)
...
-                                 true);
-
-        if(diag == HIPBLAS_DIAG_UNIT)
-        {
-            make_unit_diagonal(uplo, (T*)hA[b], lda, K);
-        }
+        make_unit_diagonal(uplo, hA);
+    }
 
+    for(int b = 0; b < batch_count; b++)
```
Comment (daineAMD) [inline clients/include/blas3/testing_trsm_batched.hpp]: can be int64_t

File: clients/include/solver/testing_getrf_batched.hpp [hunk 6]
```diff
@@ -48,9 +48,10 @@ void testing_getrf_batched_bad_arg(const Arguments& arg)
     int64_t            lda         = 102;
     int64_t            batch_count = 2;
     int64_t            A_size      = N * lda;
-    int64_t            Ipiv_size   = std::min(M, N) * batch_count;
+    int64_t            Ipiv_size   = std::min(M, N);
 
-    device_batch_vector<T> dA(A_size, 1, batch_count);
+    // Allocate device memory
+    device_batch_matrix<T> dA(M, N, lda, batch_count);
     device_vector<int>     dIpiv(Ipiv_size);
```
Comment (TorreZuk) [inline clients/include/solver/testing_getrf_batched.hpp]: Isn't the ipiv_size supposed to be * batch_count?     Not sure why they differ in style from bad_arg to normal test either.

Comment (NaveenElumalaiAMD) [reply clients/include/solver/testing_getrf_batched.hpp]: Yes it is, will change it. I think they don't differ in the normal test as it is still `std::min(M, N) * batch_count`.

File: clients/include/blas3/testing_hemm.hpp [hunk 7]
```diff
@@ -210,14 +211,13 @@ void testing_hemm(const Arguments& arg)
...
-    int64_t K    = (side == HIPBLAS_SIDE_LEFT ? M : N);
+    size_t dim_A = (side == HIPBLAS_SIDE_LEFT ? M : N);
 
     hipblasLocalHandle handle(arg);
 
     // check here to prevent undefined memory allocation error
-    bool invalid_size = M < 0 || N < 0 || ldc < M || ldb < M || lda < K;
-    if(invalid_size || !M || !N)
+    bool invalid_size = M < 0 || N < 0 || ldc < M || ldb < M || lda < dim_A;
+    if(M < 0 || N < 0 || ldc < M || ldb < M || lda < dim_A)
```
Comment (daineAMD) [inline clients/include/blas3/testing_hemm.hpp]: This is just the same as `if(invalid_size)` then if we're having !M and !N pass through.

**Review by daineAMD — APPROVED** [review]

File: clients/include/blas3/testing_hemm.hpp (hunk 7 above)
Comment (NaveenElumalaiAMD) [reply clients/include/blas3/testing_hemm.hpp]: Good catch, I will change it.

**Pull request merged**